### PR TITLE
feat: create distributed Mito2 table

### PR DESCRIPTION
**Pull request by MichaelScofield**
I hereby agree to the terms of the [GreptimeDB CLA](https://gist.github.com/xtang/6378857777706e568c1949c7578592cc)

## What's changed and what's your intention?

Create distributed Mito2 table.

If "`engine = mito2`" is in the create table SQL, create the Mito2 table. Otherwise create the original Mito table.

Note that since the Mito2 region is not able to be actually created on Datanodes now, there are no complete create Mito2 table tests. I only mock the region server's grpc interfaces in the unit tests. Once the region server is ready on Datanode, it's easy to really test it.

This PR also ships with a region server client.

## Checklist

- [ ]  I have written the necessary rustdoc comments.
- [ ]  I have added the necessary unit tests and integration tests.

## Refer to a related PR or issue link (optional)

#2213

### CODE REVIEW

File: src/meta-srv/src/procedure/create_table.rs [hunk 1]
```diff
@@ -99,17 +111,117 @@ impl CreateTableProcedure {
...
+        let mut primary_key = Vec::with_capacity(create_table_expr.primary_keys.len());
+        create_table_expr
+            .column_defs
+            .iter()
+            .enumerate()
+            .for_each(|(i, c)| {
+                let semantic_type = if create_table_expr.time_index == c.name {
+                    SemanticType::Timestamp
+                } else if create_table_expr.primary_keys.contains(&c.name) {
+                    primary_key.push(i as u32);
```
Comment (evenyag) [inline src/meta-srv/src/procedure/create_table.rs:136]: Does this respect the column order of the primary key? e.g. We have `PRIMARY_KEY(b, a)`. but in `column_defs` we store `column a, column b`.

File: src/client/src/region.rs [hunk 2]
```diff
@@ -0,0 +1,183 @@
...
+
+    #[builder(setter(strip_option), default)]
+    span_id: Option<u64>,
+
+    client: Client,
+}
+
+impl RegionClient {
+    pub async fn handle(&self, request: PbRegionRequest) -> Result<AffectedRows> {
+        let request_type = request.as_ref().to_string();
```
[the remark this answers is not in the document]
Comment (MichaelScofield) [reply src/client/src/region.rs:54]: Actually the request has derived the `strum_macros::AsRefStr`, only the enum name is "to_string"ed. See the tests here: https://github.com/GreptimeTeam/greptime-proto/pull/87/files

**Comment by MichaelScofield** [comment]
@evenyag PTAL

File: src/meta-srv/src/procedure/create_table.rs (hunk 1 above)
Comment (WenyXu) [reply src/meta-srv/src/procedure/create_table.rs:136]: We should respect the user-defined primary key orders. The following expression should have a different primary key.
`create table foo (a int, b int, primary key(a, b))`  -> (a, b)
`create table bar (a int, b int, primary key(b, a))` -> (b, a)
If we store the primary key like `a + b` in an ordered tree structure. For `(a, b)`, We  have indexed search capabilities on `(a), (a, b)`,  For `(b, a)` We have indexed search capabilities on `(b), (b, a)`.

**Comment by MichaelScofield** [comment]
@evenyag @WenyXu PTAL

**Review by WenyXu — APPROVED** [review]

File: src/client/src/region.rs (hunk 2 above)
[the remark this answers is not in the document]
Comment (evenyag) [reply src/client/src/region.rs:54]: I have merged this. You can update greptime-proto now. Is `to_string()` necessary? It returns a `&'static str` and we might don't need to convert it to string.

File: src/client/src/region.rs [hunk 3]
```diff
@@ -0,0 +1,183 @@
...
+//
+// Unless required by applicable law or agreed to in writing, software
+// distributed under the License is distributed on an "AS IS" BASIS,
+// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+// See the License for the specific language governing permissions and
+// limitations under the License.
+
+use api::v1::region::region_request::Body as PbRegionRequest;
+use api::v1::region::{self as region_pb};
+use api::v1::{AuthHeader, RequestHeader as PbRequestHeader, ResponseHeader as PbResponseHeader};
```
Comment (evenyag) [inline src/client/src/region.rs:17]: Now no structs have naming conflicts. We can import `region_request` and use `region_request::Body`.
```suggestion
use api::v1::region::region_request{self, RegionRequest};
use api::v1::{AuthHeader, RequestHeader, ResponseHeader};
```

Examples:
https://github.com/GreptimeTeam/greptimedb/pull/2269/files#diff-190ca5fa9f563aa180bb7ca434d4e3d0c3bc5d9991d8d83cc7a686218e60f157R337-R347

File: src/meta-srv/src/procedure/create_table.rs [hunk 4]
```diff
@@ -99,17 +113,126 @@ impl CreateTableProcedure {
...
+            .collect::<Vec<_>>();
+
+        let primary_key = create_table_expr
+            .primary_keys
+            .iter()
+            .map(|key| {
+                column_defs
+                    .iter()
+                    .enumerate()
+                    .find_map(|(i, c)| if &c.name == key { Some(i as u32) } else { None })
```
Comment (evenyag) [inline src/meta-srv/src/procedure/create_table.rs:159]: I remember that elements in `primary_key` field in `CreateTableRequest` are the column ids. @waynexia 

If so, we should update the document in the proto file.
```suggestion
                    .find_map(|c| if &c.name == key { Some(c.column_id) } else { None })
```

File: src/client/src/region.rs (hunk 2 above)
[the remark this answers is not in the document]
Comment (MichaelScofield) [reply src/client/src/region.rs:54]: `as_ref()` returns "&str". Besides, the `RegionRequest` will take the ownership of `request`, we can't borrow the "request.as_ref()`.

File: src/meta-srv/src/procedure/create_table.rs (hunk 4 above)
Comment (waynexia) [reply src/meta-srv/src/procedure/create_table.rs:159]: it's true

File: src/client/src/region.rs (hunk 3 above)
Comment (evenyag) [reply src/client/src/region.rs:17]: Alias is confusing and sometimes it is hard for me to search them, distinguish them, and map them to the correct pb defines.

e.g. `PbRegionRequest`, `region_pb::RegionRequest`.......

File: src/client/src/region.rs [hunk 5]
```diff
@@ -0,0 +1,182 @@
...
+    authorization: Option<AuthHeader>,
+
+    #[builder(setter(into), default)]
+    dbname: String,
+
+    #[builder(setter(strip_option), default)]
+    trace_id: Option<u64>,
+
+    #[builder(setter(strip_option), default)]
+    span_id: Option<u64>,
```
Comment (waynexia) [inline src/client/src/region.rs:46]: FWIW, this is not suitable to be a field. This is misleading. How about putting these in fn param as they should change for each fn call?

Comment (MichaelScofield) [reply src/client/src/region.rs]: Add a todo here for myself, will resolve it when we have them.

**Comment by MichaelScofield** [comment]
@waynexia PTAL

**Review by waynexia — APPROVED** [review]

**Pull request merged**